synapse/handlers/message.py

```diff
@@ -1455,7 +1457,13 @@ async def _persist_event(
                     raise
                 stream_id = result["stream_id"]
                 event_id = result["event_id"]
-                if event_id != event.event_id:
+
+                # If we batch persisted events we return the last persisted event, otherwise
+                # we return the one event that was persisted
+                event, _ = events_and_context[-1]
+
+                # We don't worry about de-duplicating batch persisted events
+                if event_id != event.event_id and len(events_and_context) == 1:
```

Comment:
TBH I don't think its worth gating this behind `len(events_and_context) == 1`. It just makes describing the behaviour harder than.